### PR TITLE
RSS-ECOMM-3-06/07 Interactive product cards, Redirect to detailed product page

### LINKED ISSUE #27: Interactive Product Cards
## Related issues in RS task

- [x] [Issue RSS-ECOMM-3_06: Interactive Product Cards](https://github.com/rolling-scopes-school/tasks/blob/master/tasks/eCommerce-Application/Sprints/Sprint3/RSS-ECOMM-3_06.md)
- [x] [Issue RSS-ECOMM-3_07: Redirect to Detailed Product Page](https://github.com/rolling-scopes-school/tasks/blob/master/tasks/eCommerce-Application/Sprints/Sprint3/RSS-ECOMM-3_07.md)

### PR DESCRIPTION
**Pull request by makarovaiuliia**
## Overview
Product card was designed and implemented in code. With the usage of `Link` component the redirection to the detailed product page was added.

**What has been done:**
- Component `Card`: design and code
- Redirection on detailed product page

**Issue(s) closed:**
issue #27 is connected to this PR

Technical task issues:
- [x] [Issue RSS-ECOMM-3_06: Interactive Product Cards](https://github.com/rolling-scopes-school/tasks/blob/master/tasks/eCommerce-Application/Sprints/Sprint3/RSS-ECOMM-3_06.md)
- [x] [Issue RSS-ECOMM-3_07: Redirect to Detailed Product Page](https://github.com/rolling-scopes-school/tasks/blob/master/tasks/eCommerce-Application/Sprints/Sprint3/RSS-ECOMM-3_07.md)

## Type of Change
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## What Reviewers Should Know
- A Component `Card` you can see on the page `/product`, it's a temporal place for it. Then It will be placed inside the catalog.

## Screen recording
https://github.com/makarovaiuliia/kickflip/assets/144937274/a8856730-909b-40f6-905b-9ca1f02f95e9

## Checklist:
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

File: kickFlip-project/src/components/app/app.tsx [hunk 1]
```diff
@@ -32,6 +38,9 @@ function App() {
             <Route path="/" element={<BasicLayoutPage />}>
                 <Route index element={<HomePage />} />
                 <Route path="products" element={<ProductsPage />} />
+                <Route path="product" element={<Card productInfo={mockData} />} />
+                {/* TODO: Вставить сюда элемент, который является страницей продукта */}
+                {/* <Route path="products/:id" element={} /> */}
```
Comment (SvetaIlina) [inline kickFlip-project/src/components/app/app.tsx:43]: у нас же есть slug, может его нужно в путь прописывать, а не id?

Comment (SvetaIlina) [inline kickFlip-project/src/components/app/app.tsx]: давай попробуем добавить slug

Comment (makarovaiuliia) [reply kickFlip-project/src/components/app/app.tsx]: Добавила

Comment (SvetaIlina) [reply kickFlip-project/src/components/app/app.tsx]: супер!

File: kickFlip-project/src/utils/utils.ts [hunk 2]
```diff
@@ -65,3 +65,38 @@ export const createBasicAuthToken = (clientId: string, clientSecret: string): st
...
+
+    const processVariant = (variant: Product) => {
+        const colorAttr = variant.attributes.find((attr) => attr.name === 'color');
+        if (colorAttr) {
+            const color = colorAttr.value;
+            if (!colorImagesMap[color]) {
+                colorImagesMap[color] = [];
+            }
+            variant.images.forEach((image) => {
+                if (colorImagesMap[color].length < 2) {
```
Comment (SvetaIlina) [inline kickFlip-project/src/utils/utils.ts]: magic number

File: kickFlip-project/src/utils/utils.ts [hunk 3]
```diff
@@ -65,3 +65,38 @@ export const createBasicAuthToken = (clientId: string, clientSecret: string): st
...
+
+    const { masterVariant } = data.masterData.current;
+    processVariant(masterVariant);
+
+    data.masterData.current.variants.forEach((variant) => {
+        processVariant(variant);
+    });
+
+    Object.values(colorImagesMap).forEach((images) => {
+        if (images.length === 2) {
```
Comment (SvetaIlina) [inline kickFlip-project/src/utils/utils.ts]: и тут)

File: kickFlip-project/src/utils/utils.ts [hunk 4]
```diff
@@ -65,3 +65,38 @@ export const createBasicAuthToken = (clientId: string, clientSecret: string): st
     const token = `${clientId}:${clientSecret}`;
     return btoa(token);
 };
+
+export const getImageFromEachColor = (data: ProductResponse) => {
+    const colorImagesMap: { [key: string]: string[] } = {};
+    const imagePairs: string[][] = [];
+
+    const processVariant = (variant: Product) => {
```
Comment (SvetaIlina) [inline kickFlip-project/src/utils/utils.ts:72]: и я бы вынесла эту функцию отдельно, чтобы ее можно было переиспользовать в другом месте.

Comment (makarovaiuliia) [reply kickFlip-project/src/utils/utils.ts:72]: Сделала. Сама функция процессинга одного варианта осталась внутри, но теперь есть функция, которая выдет map со всеми цветами и всеми фотками

File: kickFlip-project/src/utils/utils.ts (hunk 3 above)
Comment (makarovaiuliia) [reply kickFlip-project/src/utils/utils.ts]: Убрала, теперь выдает все фотки, а не только как мне надо было - две)

**Review by SvetaIlina — APPROVED** [review]

**Review by Pavel-Kuvshinov — APPROVED** [review]
Code is perfect

**Pull request merged**